src/Database/LSMTree/Internal/Serialise/RawBytes.hs

```diff
@@ -0,0 +1,193 @@
+--
+-- The /top/ corresponds to the most significant bit (big-endian).
+--
+-- PRECONDITION: @n >= 0 && n <= 16. We can slice out at most 16 bits,
+-- all bits beyond that are truncated.
+--
+-- PRECONDITION: The byte-size of the raw bytes should be at least 2 bytes.
+--
+-- TODO: optimisation ideas: use unsafe shift/byteswap primops, look at GHC
+-- core, find other opportunities for using primops.
+--
+topBits16 :: Int -> RawBytes -> Word16
+topBits16 n rb@(RawBytes (P.Vector (I# off#) _size (ByteArray k#))) =
+    assert (sizeofRawBytes rb >= 2) $ shiftR w16 (16 - n)
+  where
+    w16 = toWord16 (indexWord8ArrayAsWord16# k# off#)
+
+toWord16 :: Word16# -> Word16
+#if WORDS_BIGENDIAN
+toWord16 = W16#
+#else
+toWord16 x# = byteSwap16 (W16# x#)
+#endif
+
+-- | @'sliceBits32' off rb@ slices from the raw bytes @rb@ a string of @32@
+-- bits, starting at the @0@-based offset @off@. Returns the string of bits as a
+-- 'Word32'.
+--
+-- Offsets are counted in bits from the /top/: offset @0@ corresponds to the
+-- most significant bit (big-endian).
+--
+-- PRECONDITION: The raw bytes should be large enough that we can slice out 4
+-- bytes after the bit-offset @off@, since we can only slice out bits that are
+-- within the bounds of the byte array.
+--
+-- TODO: optimisation ideas: use unsafe shift/byteswap primops, look at GHC
+-- core, find other opportunities for using primops.
+--
+sliceBits32 :: Int -> RawBytes -> Word32
+sliceBits32 off@(I# off1#) rb@(RawBytes (P.Vector (I# off2#) _size (ByteArray ba#)))
+    | 0# <- r#
+    = assert (off + 32 <= 8 * sizeofRawBytes rb) $
+      toWord32 (indexWord8ArrayAsWord32# ba# q#)
+    | otherwise
+    = assert (off + 32 <= 8 * sizeofRawBytes rb) $
+        toWord32 (indexWord8ArrayAsWord32# ba# q#       ) `shiftL` r
+      + w8w32#   (indexWord8Array#         ba# (q# +# 4#)) `shiftR` (8 - r)
+  where
+    !(# q0#, r# #) = quotRemInt# off1# 8#
+    !q#            = q0# +# off2#
+    r              = I# r#
+    -- No need for byteswapping here
+    w8w32# x#     = W32# (wordToWord32# (word8ToWord# x#))
+
+toWord32 :: Word32# -> Word32
+#if WORDS_BIGENDIAN
+toWord32 = W32#
+#else
+toWord32 x# = byteSwap32 (W32# x#)
+#endif
+
+sizeofRawBytes :: RawBytes -> Int
+sizeofRawBytes (RawBytes pvec) = P.length pvec
+
+concat :: [RawBytes] -> RawBytes
+concat rbs = RawBytes (P.concat (fmap toVector rbs))
+
+{-------------------------------------------------------------------------------
+  @bytestring@ utils
+-------------------------------------------------------------------------------}
+
+-- | \( O(1) \) conversion from a strict bytestring to raw bytes.
+unsafeFromByteString :: BS.ByteString -> RawBytes
+unsafeFromByteString (BS.Internal.BS (GHC.ForeignPtr _ contents) n) =
+    case contents of
+      -- Strict bytestrings are allocated using 'mallocPlainForeignPtrBytes', so
+      -- we are expecting a 'PlainPtr' here.
+      PlainPtr mba# -> RawBytes $ P.Vector 0 n $ ByteArray (unsafeCoerce# mba#)
```

Comment:
Yes, one can use `unsafeCoerce#` but it's a very big hammer.

A somewhat smaller hammer is using `unsafeFreezeByteArray#` in combination with `realWorld#`, i.e.
```
PlainPtr mba# -> case unsafeFreezeByteArray mba# realWorld# of
                   (# _, ba# #) -> RawBytes (P.Vector 0 n (ByteArray ba#))
```